### PR TITLE
Flatten Failure Metadata

### PR DESCRIPTION
**Pull request by cjllanwarne**

### CODE REVIEW

**Comment by geoffjentry** [comment]
Make sure @ansingh7115 takes a look at the new output if she hasn't already

File: services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala [hunk 1]
```diff
@@ -74,4 +77,94 @@ class MetadataServiceSpec extends FlatSpec with Matchers with TableDrivenPropert
     }
   }
 
+  // For the metadata event tests!
+  val eventKeyRegex = "path\\:to\\:failures\\[([0-9]*)\\]\\:message".r
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala]: should that `*` be a `+`?

File: services/src/main/scala/cromwell/services/metadata/MetadataService.scala [hunk 2]
```diff
@@ -129,10 +131,12 @@ object MetadataService {
   def throwableToMetadataEvents(metadataKey: MetadataKey, t: Throwable): List[MetadataEvent] = {
     t match {
       case aggregation: ThrowableAggregation =>
-        aggregation.errorMessages.toList map { message => MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}:message"), MetadataValue(s"${aggregation.exceptionContext}: $message")) }
+        val message = List(MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}[${Random.nextInt(Int.MaxValue)}]:message"), MetadataValue(aggregation.exceptionContext)))
```
Comment (mcovarr) [inline services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: I realize this is somewhat a refactoring of what was here previously, but the space of nonnegative `Int`s is kind of limited for a hash...

**Comment by cjllanwarne** [comment]
Approved in person. Migration options TBD

File: services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala [hunk 3]
```diff
@@ -74,4 +77,94 @@ class MetadataServiceSpec extends FlatSpec with Matchers with TableDrivenPropert
...
+      validateEvent(events(3), workflowId, innerCauseMsg),
+      validateEvent(events(4), workflowId, causeMsg2),
+      validateEvent(events(5), workflowId, causeMsg3)
+    )
+    // Make sure all IDs are unique:
+    errorIds.toSet.size should be(6)
+  }
+
+  def validateEvent(event: MetadataEvent, workflowId: WorkflowId, message: String) = event match {
+    case MetadataEvent(k, Some(MetadataValue(v, _)), _) =>
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala]: if this doesn't match at all there's no error?

File: services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala [hunk 4]
```diff
@@ -74,4 +77,94 @@ class MetadataServiceSpec extends FlatSpec with Matchers with TableDrivenPropert
...
+      k.workflowId should be(workflowId)
+      val messageIndex = k.key match {
+        case eventKeyRegex(x) => x
+        case _ => fail("Unexpected failure key format: " + k.key)
+      }
+      v should be(message)
```
Comment (mcovarr) [inline services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala]: the `v should...` would probably make more sense under the `k.workflowId should...` and then you can return the `k.key match {...}` expression directly

**Comment by mcovarr** [comment]
:+1:

File: services/src/main/scala/cromwell/services/metadata/MetadataService.scala [hunk 5]
```diff
@@ -129,10 +131,12 @@ object MetadataService {
...
+        val innards = aggregation.throwables.toList flatMap { innard => throwableToMetadataEvents(metadataKey, innard) }
+        message ++ innards
       case other =>
-        val message = List(MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}:message"), MetadataValue(t.getMessage)))
-        val cause = Option(t.getCause) map { cause => throwableToMetadataEvents(metadataKey.copy(key = s"${metadataKey.key}:causedBy"), cause) } getOrElse List.empty
+        val message = List(MetadataEvent(metadataKey.copy(key = s"${metadataKey.key}[${Random.nextInt(Int.MaxValue)}]:message"), MetadataValue(t.getMessage)))
+        val cause = Option(t.getCause) map { cause => throwableToMetadataEvents(metadataKey, cause) } getOrElse List.empty
```
Comment (kshakir) [inline services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: Injecting the `Random.nextInt` here _inside_ `throwableToMetadataEvents` on 138 separates the exception from the recursive underlying cause on 139. IMHO that makes debugging much more confusing. These are _not_ separate events, and should be tied together.

Btw, will tech talk later, but as far as I can tell we only have **one** active json schema for exceptions, the one mentioned in the [pseudo-CFG](https://github.com/broadinstitute/cromwell/issues/2037#issuecomment-282800499).

The inactive json schema that may have triggered this whole issue appears to be an artifact of the [migration](https://github.com/broadinstitute/cromwell/blob/509ff2dc2bc2208546a76307791af05816ec83f5/database/migration/src/main/scala/cromwell/database/migration/metadata/table/FailureEventTableMigration.scala#L23) [code](https://github.com/broadinstitute/cromwell/blob/509ff2dc2bc2208546a76307791af05816ec83f5/database/migration/src/main/scala/cromwell/database/migration/metadata/table/FailureEventTableMigration.scala#L47) not matching the active code, deleted in red above.

I'm fine upgrading and migrating our current the json [schema](http://json-schema.org/) to something non-recursive, but would propose something along the lines of:

```json
{
  "$ref": "#/definitions/failures",
  "definitions": {
    "failures": {
      "type": "array",
      "items": {
        "$ref": "#/definitions/failure"
      }
    },
    "failure": {
      "required": ["message"],
      "properties": {
        "message": {
          "type": "string"
        },
        "causedBy": {
          "type": "array",
          "items": {
            "type": "string"
          }
        }
      }
    }
  }
}
```

This schema would change the recursive:

```json
{
  "failures": [
    {
      "message": "Failed to upload authentication file",
      "causedBy": {
        "message": "Error getting access token for service account: ",
        "causedBy": {
          "message": "connect timed out"
        }
      }
    }
  ]
}
```

to the more condensed:

```json
{
  "failures": [
    {
      "message": "Failed to upload authentication file",
      "causedBy": [
        "Error getting access token for service account: ",
        "connect timed out"
      ]
    }
  ]
}
```

and would still allow one to figure out what exceptions are connected, or not:

```json
{
  "failures": [
    {
      "message": "Internal unconnected exception with no cause by"
    },
    {
      "message": "A separate I/O exception that failed to upload authentication file",
      "causedBy": [
        "Error getting access token for service account: ",
        "connect timed out"
      ]
    }
  ]
}
```

Also fine with re-migrating the old data to bring it up to a current json schema implementation.

Comment (cjllanwarne) [inline services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: @helgridly @ansingh7115 would you like to join our tech talk, at the appropriate time?

File: services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala (hunk 1 above)
Comment (cjllanwarne) [inline services/src/test/scala/cromwell/services/metadata/MetadataServiceSpec.scala]: Yeah... I mean it's just a case of where an exception is thrown in a test, but... yeah.

File: services/src/main/scala/cromwell/services/metadata/MetadataService.scala (hunk 5 above)
Comment (helgridly) [inline services/src/main/scala/cromwell/services/metadata/MetadataService.scala]: Yes. Short version: this seems fine. For now we're just looking at the message field of each object anyway, so you can throw whatever other <strike>junk</strike> useful information in there you'd like.

**Comment by cjllanwarne** [comment]
Closing this PR. The branch may or may not be repurposed.

**Pull request closed without merging**